tests/test_modeling_common.py

```diff
@@ -575,20 +575,69 @@ def test_training_gradient_checkpointing(self):
                 or not model_class.supports_gradient_checkpointing
             ):
                 continue
+            # Scenario - 1 default behaviour
             model = model_class(config)
             model.to(torch_device)
             model.gradient_checkpointing_enable()
             model.train()
+
+            # unfreeze additional layers
+            for p in model.parameters():
+                p.requires_grad_(True)
+
+            optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
+
             inputs = self._prepare_for_class(inputs_dict, model_class, return_labels=True)
             loss = model(**inputs).loss
             loss.backward()
+            optimizer.step()
 
-            model.gradient_checkpointing_disable()
+            for k, v in model.named_parameters():
+                if v.requires_grad:
+                    self.assertTrue(v.grad is not None, f"{k} in {model_class.__name__} has no gradient!")
+
+            # Scenario - 2 with `use_reentrant=True`
+            model = model_class(config)
+            model.to(torch_device)
             model.gradient_checkpointing_enable(gradient_checkpointing_kwargs={"use_reentrant": True})
             model.train()
+
+            # unfreeze additional layers
+            for p in model.parameters():
+                p.requires_grad_(True)
+
+            optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
+
             inputs = self._prepare_for_class(inputs_dict, model_class, return_labels=True)
             loss = model(**inputs).loss
             loss.backward()
+            optimizer.step()
+
+            for k, v in model.named_parameters():
+                if v.requires_grad:
+                    self.assertTrue(v.grad is not None, f"{k} in {model_class.__name__} has no gradient!")
+
+            # Scenario - 3 with `use_reentrant=False`
```

Comment:
might be better to split this test to 3 tests (we can define a method that has the common part between these 3 tests)

Reply:
+1